mne/io/edf/edf.py

```diff
@@ -227,61 +212,61 @@ def _read_segment(self, start=0, stop=None, sel=None, verbose=None,
                         else:
                             warnings.warn('Interpolating stim channel.'
                                           ' Events may jitter.')
-                            oldrange = np.linspace(0, 1, samp + 1,
-                                                   True)
-                            newrange = np.linspace(0, 1, max_samp,
-                                                   False)
+                            oldrange = np.linspace(0, 1, samp + 1, True)
+                            newrange = np.linspace(0, 1, max_samp, False)
                             ch_data = interp1d(
                                 oldrange, np.append(ch_data, 0),
                                 kind='zero')(newrange)
                     elif samp != max_samp:
                         ch_data = resample(x=ch_data, up=max_samp, down=samp,
                                            npad=0)
-                    data[j] = ch_data
-                start_pt = int(max_samp * i)
-                stop_pt = int(start_pt + max_samp)
-                datas[:, start_pt:stop_pt] = data
-        datas *= gains.T
-        datas += offsets
+                    this_data[count, :] = ch_data
+                    count += 1
+                s_off = start - blockstart if bi == 0 else 0
+                e_off = stop_pt - stop + blockstart if bi == n_blk - 1 else 0
+                assert s_off >= 0
+                assert e_off >= 0
+                time_slice = slice(start_pt + (s_off - start + blockstart),
+                                   stop_pt - (e_off + start - blockstart),
+                                   None)
+                data[:, time_slice] = this_data[:, s_off:buffer_size - e_off]
```

Comment:
`e_off` is the difference between where the chunk of data being read in ends (which always has length `buffer_size`) and where the last point of data we actually need to read is. Perhaps I'll make a diagram.

Incidentally this is part 1 of where the big savings comes in. The idea is to only allocate `data` once, and fill it. This means (1) only inserting data from channels we actually need (see `continue` line above) and (2) only storing data from the time slices from each block that we actually need. You need to worry about adjusting the start point of the first block that is read, the end point of the last block that is read, and where these chunks actually go in your data, to get that second part right.